RFCs/0043-Fn::FindInMap enhancements (default value and intrinsic functions).md

```diff
@@ -0,0 +1,571 @@
+
+We will add support for using other intrinsic functions, including `Fn::Select`, `Fn::Join`, `Fn::Sub`, and `Fn::If` within `Fn::FindInMap`. We will also add an optional default value for `Fn::FindInMap`, which will be the output of the `Fn::FindInMap` if the mapping is not present. The first three parameters of `Fn::FindInMap` will remain the same as before, while the optional fourth parameter `DefaultValue` will be of type `String` or `List`.
+
+# Motivation
+
+## Use of other intrinsic functions:
+
+CloudFormation users may want to use intrinsic functions such as `Fn::Select`, `Fn::Join`, `Fn::Sub`, and `Fn::If` within `Fn::FindInMap`. While CloudFormation users can currently nest `Fn::Ref` and `Fn::FindInMap` in `Fn::FindInMap`, more function support has been requested to give CloudFormation users more flexibility when using `Fn::FindInMap`.
+
+* Reference:
+    * https://github.com/aws-cloudformation/cfn-language-discussion/issues/73
+    * https://github.com/aws-cloudformation/cfn-language-discussion/issues/79
+
+## Default Value:
+
+CloudFormation users may want to reuse their templates across many different use cases, such as different regions. Mappings and `Fn::FindInMap` allow users to switch between variables via parameters, but if a Mapping cannot be found using the provided `MapName`, `TopLevelKey`, and `SecondLevelKey`, an error will be thrown. Users must specify all possible permutations of values in a Mapping, which can result in a significant amount of duplicate code. 
+
+An example is as follows. 
+
+### YAML
+```yaml
+Mappings:
+  DNS:
+    us:
+      dns: mypage.us.com
+      ttl: '600'
+    ca:
+      dns: mypage.us.com
+      ttl: '600'
+    mx:
+      dns: mypage.default.com
+      ttl: '300'
+    ar:
+      dns: mypage.default.com
+      ttl: '300'
+    br:
+      dns: mypage.default.com
+      ttl: '300'
+
+!FindInMap [DNS, !Ref country, dns]
+```
+
+### JSON
+```json
+{
+    "Mappings" : {
+        "DNS": {
+            "us" : {"dns" : "mypage.us.com", "ttl" : "600"},
+            "ca" : {"dns" : "mypage.us.com", "ttl" : "600"},
+            "mx" : {"dns" : "mypage.default.com", "ttl" : "300"},
+            "ar" : {"dns" : "mypage.default.com", "ttl" : "300"},
+            "br" : {"dns" : "mypage.default.com", "ttl" : "300"}
+        }
+    }
+}
+
+{ "Fn::FindInMap" : [ "DNS", { "Ref" : "country" }, "dns" ]}
+```
+
+The desired behavior would be something like this:
+
+```yaml
+!If
+- !MappingPresent [Mapping, 'foo', 'bar']
+- !FindInMap [Mapping, 'foo', 'bar']
+- 'MyDefaultValue'
+```
+
+The above example could then be simplified as follows.
+
+### YAML
+```yaml
+Mappings:
+  DNS:
+    us:
+      dns: mypage.us.com
+      ttl: '600'
+    ca:
+      dns: mypage.us.com
+      ttl: '600'
+
+!FindInMap [DNS, !Ref country, dns, mypage.default.com]
```

Comment:
Is the plan to both support the `MappingPresent` function and a new `!FindInMap` function signature? 

From the details section it seem like only the `!FindInMap` way will be included (which is fine)

~~In either case, I think a definition/explanation of `!FindInMap [Mapping, TopLevel, SecondLevel, Default]` would be useful.~~ This is included at the end

Another way to solve the same problem - with different trade-offs - is to have a special key in the mapping itself:

```
Mappings:
  DNS:
    us:
      dns: mypage.us.com
      ttl: '600'
    ca:
      dns: mypage.us.com
      ttl: '600'
    $default:
      dns: mypage.us.com
      ttl: '600'

# This automatically gets the default, so it's less explicit about what's happening
!FindInMap [DNS, !Ref country, dns]

# It could be made more explicit if needed (could be the only supported way too)
# This option could support multiple user-provided $-keys to get the best from both worlds
!FindInMap [DNS, !Ref country, dns, $default]  

```

For the described use case I think it's more useful to have only one place to define the default, otherwise, if you need to get the value more than one,  the more maintainable way of writing it would be something like

```
Mappings:
  DNS:
    us:
      dns: mypage.us.com
      ttl: '600'
    ca:
      dns: mypage.us.com
      ttl: '600'
  Defaults:
    DNS:
      dns: mypage.default.com
      ttl: 300

!FindInMap [DNS, !Ref country, dns, !FindInMap [Defaults, DNS, dns]]
!FindInMap [DNS, !Ref country, ttl, !FindInMap [Defaults, DNS, ttl]]
```
or
```
Mappings:
  DNS:
    us:
      dns: mypage.us.com
      ttl: '600'
    ca:
      dns: mypage.us.com
      ttl: '600'
    default:
      dns: mypage.default.com
      ttl: 300

!FindInMap [DNS, !Ref country, dns, !FindInMap [DNS, default, dns]]
!FindInMap [DNS, !Ref country, ttl, !FindInMap [DNS, default, ttl]]
```

Reply:
That's an interesting idea. 
```yaml
Mappings:
  DNS:
    us:
      dns: mypage.us.com
      ttl: '600'
    ca:
      dns: mypage.us.com
      ttl: '600'
  Defaults:
    DNS:
      dns: mypage.default.com
      ttl: 300
```

Given the fact we would like to support more intrinsic functions in `!FindInMap`, 
```yaml
!FindInMap [DNS, !Ref country, dns, !FindInMap [Defaults, DNS, dns]]
!FindInMap [DNS, !Ref country, ttl, !FindInMap [Defaults, DNS, ttl]]
```
Should be a valid use case.

As for setting a special key in mapping as default key - I'm a bit concerned as it may break customers who already use it as a regular key.

Initially during our internal discussion, we thought adding an optional parameter makes it easy for customers to onboard this feature - you don't need to refactor your `FindInMap` function, just append the argument list then you are good to go. But now I'm thinking, maybe `MappingPresent` can be used without the context of `FindInMap` default value: it's a conditional function so it can alter the behavior of CFN engine, e.g provision a resource conditionally. Yet I don't have any concrete use cases on top of my head to support how useful this new `MappingPresent` function would be.

Reply:
I didn't check, but I seem to remember some values being not allowed in mapping keys, so depending on the list of values there can be a way to introduce special keys without breaking existing templates

Reply:
Hi Ben, sorry for the late reply. 

> Another way to solve the same problem - with different trade-offs - is to have a special key in the mapping itself

So I did some investigation on reserving a special key as default. But I decided to rule it out because:
1. IMHO, a reserved key is less flexible compare to adding an optional `DefaultValue` parameter. Customers sometimes may need to define different default values for the same mapping under different circumstances. With `$default` key defined in mapping section, they are stuck with the pre-defined value.
2. It’s hard to define what exactly `$default` would contain: should it be a TopLevelKey or SecondLevelKey? To maintain extensibility, define $default as a SecondLevelKey without TopLevelKey seems make more sense but it would be confusing to customers because `$default` does not maintain the same structure as other entries.
3. This enhancement is delivered via `AWS::LanguageExtensions`, so the response of our transform has to be a legit CloudFormation template. If we reserve a Key without special characters, we run into risk where customers may have already define that key. If we do reserve a key with special characters, we need to remove `$default` section before returning template. This, imo is counter intuitive and does not provide the best customer experience.

> However in the examples us-east-1 is used too. There is a potential to find out the exact allowed pattern by calling aws cloudformation validate-template, but having the documentation be updated is probably better

This is a great catch! *-* is allowed in key and I'll reach out to doc team to make an update.